primitives/api/proc-macro/src/runtime_metadata.rs

```diff
@@ -0,0 +1,253 @@
+	for generic_param in generics.params.iter_mut() {
+		let syn::GenericParam::Type(ty) = generic_param else {
+			continue
+		};
+
+		// `scale_info::meta_type` requires `T: ?Sized + TypeInfo + 'static` bounds.
+		ty.bounds.push(parse_quote!(#crate_::scale_info::TypeInfo));
+		ty.bounds.push(parse_quote!('static));
+		// Default type parameters are not allowed in functions.
+		ty.eq_token = None;
+		ty.default = None;
+	}
+
+	let where_clause: Vec<_> = where_clause
+		.iter()
+		.map(|ty| quote!(#ty: #crate_::scale_info::TypeInfo + 'static))
+		.collect();
+
+	quote!(
+		#( #attrs )*
+		#[inline(always)]
+		pub fn runtime_metadata #generics () -> #crate_::metadata_ir::RuntimeApiMetadataIR
+		where #( #where_clause, )*
+		{
+			#crate_::metadata_ir::RuntimeApiMetadataIR {
+				name: #trait_name,
+				methods: #crate_::vec![ #( #methods, )* ],
+				docs: #docs,
+			}
+		}
+	)
+}
+
+/// Implement the `runtime_metadata` function on the runtime that
+/// generates the metadata for the given traits.
+///
+/// The metadata of each trait is extracted from the generic function
+/// exposed by `generate_decl_runtime_metadata`.
+pub fn generate_impl_runtime_metadata(
+	impls: &[ItemImpl],
+	crate_: &TokenStream,
+) -> Result<TokenStream> {
+	if impls.is_empty() {
+		return Ok(quote!())
+	}
+
+	// Get the name of the runtime for which the traits are implemented.
+	let runtime_name = &impls
+		.get(0)
+		.expect("Traits should contain at least one implementation; qed")
+		.self_ty;
+
+	let mut metadata = Vec::new();
+
+	for impl_ in impls {
+		let mut trait_ = extract_impl_trait(&impl_, RequireQualifiedTraitPath::Yes)?.clone();
+
+		// Implementation traits are always references with a path `impl client::Core<generics> ...`
+		// The trait name is the last segment of this path.
+		let trait_name_ident = &trait_
+			.segments
+			.last()
+			.as_ref()
+			.expect("Trait path should always contain at least one item; qed")
+			.ident;
+
+		// Extract the generics from the trait to pass to the `runtime_metadata`
+		// function on the hidden module.
+		let generics = trait_
+			.segments
+			.iter()
+			.find_map(|segment| {
+				if let syn::PathArguments::AngleBracketed(generics) = &segment.arguments {
+					Some(generics.clone())
+				} else {
+					None
+				}
+			})
+			.expect("Trait path should always contain at least one generic parameter; qed");
+
+		let mod_name = generate_runtime_mod_name_for_trait(&trait_name_ident);
+		// Get absolute path to the `runtime_decl_for_` module by replacing the last segment.
+		if let Some(segment) = trait_.segments.last_mut() {
+			*segment = parse_quote!(#mod_name);
+		}
+
+		let attrs = filter_cfg_attributes(&impl_.attrs);
+		metadata.push(quote!(
+			#( #attrs )*
+			#trait_::runtime_metadata::#generics()
+		));
+	}
+
+	Ok(quote!(
+		trait InternalImplRuntimeApis {
+			#[inline(always)]
+			fn runtime_metadata(&self) -> #crate_::vec::Vec<#crate_::metadata_ir::RuntimeApiMetadataIR> {
```

Comment:
```suggestion
			fn runtime_metadata() -> #crate_::vec::Vec<#crate_::metadata_ir::RuntimeApiMetadataIR> {
```

No need to take any kind of `self`.